### PR TITLE
Feature: add comment edit date

### PR DESCRIPTION
**Pull request by MatthieuSLR9**
# Description

Added the time comments where edited which is shown in the UI in edit notes. Also added the name of the person who wrote the comment in the comment class which is also displayed in the UI.

Fixes # (issue)

# How Has This Been Tested?

### Unit Tests
  I tested the conversion of Comments from String to the Comment class and from it's Comment Class to it's String by updating the UI tests in NoteRepositoryFirestoreTest.
### UI Tests
I tested that when a comment is added and edited it is shown in the label text that it has been edited.
### End to end Tests
none
# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged in downstream modules, no conflicts with main

### CODE REVIEW

File: app/src/main/java/com/github/onlynotesswent/model/note/Note.kt [hunk 1]
```diff
@@ -97,11 +97,20 @@ data class Note(
        */
       fun addComment(
           userId: String,
+          userName: String,
```
Comment (santiagoarayaramos) [inline app/src/main/java/com/github/onlynotesswent/model/note/Note.kt:100]: Update Documentation to include new parameter

File: app/src/main/java/com/github/onlynotesswent/model/note/Note.kt [hunk 2]
```diff
@@ -121,7 +130,13 @@ data class Note(
         return CommentCollection(
             mutableCommentsList.map {
               if (it.commentId == commentId)
-                  Comment(generateId(it.userId, content), it.userId, content)
+                  Comment(
```
Comment (santiagoarayaramos) [inline app/src/main/java/com/github/onlynotesswent/model/note/Note.kt:133]: Seems wasteful to use a map, why not retrieve the index of the comment with `.indexOfFirst(it.commentId == commentId)`, then modify only that element. It is possible this is equivalent based on how these functions are implemented, but still seems a clearer implementation

File: app/src/main/java/com/github/onlynotesswent/model/note/Note.kt [hunk 3]
```diff
@@ -150,5 +165,17 @@ data class Note(
    * @property userId The unique identifier of the user who posted the comment.
    * @property content The text content of the comment.
    */
-  data class Comment(val commentId: String, val userId: String, val content: String)
+  data class Comment(
```
Comment (santiagoarayaramos) [inline app/src/main/java/com/github/onlynotesswent/model/note/Note.kt:168]: Update documentation to reflect the added properties

File: app/src/main/java/com/github/onlynotesswent/model/note/Note.kt [hunk 4]
```diff
@@ -150,5 +165,17 @@ data class Note(
...
+      val commentId: String,
+      val userId: String,
+      val userName: String,
+      val content: String,
+      val creationDate: Timestamp,
+      val editedDate: Timestamp
+  ) {
+    // Method to check if creationDate is equal to editedDate
+    fun isUnedited(): Boolean {
```
Comment (santiagoarayaramos) [inline app/src/main/java/com/github/onlynotesswent/model/note/Note.kt:177]: As this function designed to be public (used in EditNote), document it with /** */

File: app/src/main/java/com/github/onlynotesswent/model/note/NoteRepositoryFirestore.kt [hunk 5]
```diff
@@ -26,6 +26,47 @@ class NoteRepositoryFirestore(private val db: FirebaseFirestore) : NoteRepositor
...
+   * @throws IndexOutOfBoundsException if the comment snapshot is improperly formatted and does not
+   *   contain the expected number of fields.
+   */
+  private fun convertCommentStringToComment(commentSnapshot: String): Note.Comment {
+    val commentValues = commentSnapshot.split(commentDelimiter)
+    return Note.Comment(
+        commentValues[0],
+        userId = commentValues[1],
+        userName = commentValues[2],
+        content = commentValues[2],
```
Comment (santiagoarayaramos) [inline app/src/main/java/com/github/onlynotesswent/model/note/NoteRepositoryFirestore.kt:66]: I believe this is an error, should be` content = commentValues[3],`, and adapt the following indexes accordingly

File: app/src/main/java/com/github/onlynotesswent/ui/overview/EditNote.kt [hunk 6]
```diff
@@ -207,7 +209,10 @@ fun EditNoteScreen(
                   onClick = {
                     updatedComments =
                         Note.CommentCollection.addComment(
-                            userViewModel.currentUser.value?.uid ?: "1", "", updatedComments)
```
Comment (santiagoarayaramos) [inline app/src/main/java/com/github/onlynotesswent/ui/overview/EditNote.kt:210]: This implementation is weird. See larger comment on a bigger refactor of CommentCollection class, which will require changes here.

**Review by santiagoarayaramos — CHANGES_REQUESTED** [review]
Good work! Some nice modularity was added, and new features are good. Some documentation changes required (including changing the doc for the CommentCollection class that doesn't correctly represent what it does), and some minor bug fixing.

Most importantly, the implementation of CommentCollection seems bizarre, I recommend removing the functions currently defined in the companion class (making them static functions, in which you currently pass the class instance as a parameter), and moving them to be defined in the main class. In fact, I don't think the companion object serves any purpose (what is stated in documentation is incorrect). 
This refactor would imply removing the parameter `commentCollection` from functions, instead directly using the `commentsList` property of the collection. The returns for functions should also be removed (or set to Unit if you want to explicitly set them). Finally the use of this class in EditNotes should be modified to work with this new implementation. Don't forget to update the docs too.

Though this is not a problem introduced by this PR, I feel it is consistent and prudent to fix it now (if you agree with the proposed changes)

**Comment by MatthieuSLR9** [comment]
> Good work! Some nice modularity was added, and new features are good. Some documentation changes required (including changing the doc for the CommentCollection class that doesn't correctly represent what it does), and some minor bug fixing.
> 
> Most importantly, the implementation of CommentCollection seems bizarre, I recommend removing the functions currently defined in the companion class (making them static functions, in which you currently pass the class instance as a parameter), and moving them to be defined in the main class. In fact, I don't think the companion object serves any purpose (what is stated in documentation is incorrect). This refactor would imply removing the parameter `commentCollection` from functions, instead directly using the `commentsList` property of the collection. The returns for functions should also be removed (or set to Unit if you want to explicitly set them). Finally the use of this class in EditNotes should be modified to work with this new implementation. Don't forget to update the docs too.
> 
> Though this is not a problem introduced by this PR, I feel it is consistent and prudent to fix it now (if you agree with the proposed changes)

I have to disagree with changing the comment Collection class as the goal of using a companion object and to pass the Comment Collection in  the methods was to make the Comment Collection immutable with  the only way to modify it is by producing a new one using the companion object static methods.

File: app/src/main/java/com/github/onlynotesswent/ui/overview/EditNote.kt (hunk 6 above)
Comment (santiagoarayaramos) [reply app/src/main/java/com/github/onlynotesswent/ui/overview/EditNote.kt:210]: To be discussed later

File: app/src/main/java/com/github/onlynotesswent/model/note/Note.kt (hunk 2 above)
Comment (santiagoarayaramos) [reply app/src/main/java/com/github/onlynotesswent/model/note/Note.kt:134]: Implementation decision to be discussed later, need to consider possible collisions with commentID and what to do with it.

**Review by santiagoarayaramos — APPROVED** [review]
Some changes to be discussed later, but requested changes introduced in this pr addressed, code is clean. Good work!

**Pull request merged**